### PR TITLE
K8SPXC-673: Add PMM client sidecar to HAProxy statefulset

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPXC-673](https://jira.percona.com/browse/K8SPXC-673) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/1378af4dd10b81e6d2f914c4f8568637544596d3
image: `perconalab/percona-xtradb-cluster-operator:PR-851-1378af4d`

File: pkg/pxc/app/pmm.go [hunk 1]
```diff
@@ -161,14 +161,16 @@ func pmmAgentEnvs(pmmServerHost, pmmServerUser, secrets string) []corev1.EnvVar
...
+		pmmServerArgs += "$(DB_ARGS) "
+	}
+	if dbType != "haproxy" {
+		pmmServerArgs += "--host=$(POD_NAME) --port=$(DB_PORT)"
 	}
 	return []corev1.EnvVar{
 		{
 			Name:  "PMM_AGENT_PRERUN_SCRIPT",
-			Value: "pmm-admin status --wait=10s;\npmm-admin add $(DB_TYPE)" + pmmServerArgs + ";\npmm-admin annotate --service-name=$(PMM_AGENT_SETUP_NODE_NAME) 'Service restarted'",
+			Value: "pmm-admin status --wait=10s;\npmm-admin add $(DB_TYPE)" + pmmServerArgs + " $(PMM_AGENT_SETUP_NODE_NAME);\npmm-admin annotate --service-name=$(PMM_AGENT_SETUP_NODE_NAME) 'Service restarted'",
```
Comment (delgod) [inline pkg/pxc/app/pmm.go:173]: maybe missed option
```suggestion
			Value: "pmm-admin status --wait=10s;\npmm-admin add $(DB_TYPE)" + pmmServerArgs + " --service-name=$(PMM_AGENT_SETUP_NODE_NAME);\npmm-admin annotate --service-name=$(PMM_AGENT_SETUP_NODE_NAME) 'Service restarted'",
```

File: pkg/pxc/app/statefulset/haproxy.go [hunk 2]
```diff
@@ -258,7 +258,76 @@ func (c *HAProxy) LogCollectorContainer(spec *api.LogCollectorSpec, logPsecrets
 }
 
 func (c *HAProxy) PMMContainer(spec *api.PMMSpec, secrets string, cr *api.PerconaXtraDBCluster) (*corev1.Container, error) {
-	return nil, nil
```
Comment (delgod) [inline pkg/pxc/app/statefulset/haproxy.go:261]: should be added only since 1.9.0

File: pkg/pxc/app/statefulset/haproxy.go [hunk 3]
```diff
@@ -258,7 +258,76 @@ func (c *HAProxy) LogCollectorContainer(spec *api.LogCollectorSpec, logPsecrets
...
+
+	ct.Env = append(ct.Env, pmmEnvs...)
+	ct.Env = append(ct.Env, dbEnvs...)
+	res, err := app.CreateResources(spec.Resources)
+	if err != nil {
+		return nil, fmt.Errorf("create resources error: %v", err)
+	}
+	ct.Resources = res
+
+	if cr.CompareVersionWith("1.7.0") >= 0 {
```
Comment (delgod) [inline pkg/pxc/app/statefulset/haproxy.go:313]: whole block should be added only if version >= 1.9.0
it has no sense to add part of block only for >= 1.7.0

File: pkg/pxc/app/pmm.go [hunk 4]
```diff
@@ -90,7 +90,7 @@ func pmmAgentEnvs(pmmServerHost, pmmServerUser, secrets string) []corev1.EnvVar
 			},
 		},
 		{
-			Name: "POD_NAMESPASE",
+			Name: "POD_NAMESPACE",
```
Comment (delgod) [inline pkg/pxc/app/pmm.go:93]: such renmae should be only if version >= 1.9.0

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/pxc/app/pmm.go (hunk 1 above)
Comment (egegunes) [reply pkg/pxc/app/pmm.go:173]: pmm-admin doesn't have `--service-name` option for `haproxy` but it seems all `add` subcommands accept the service name as a positional argument

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| storage | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| haproxy | failed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| security-context | passed |
| limits | failed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/1378af4dd10b81e6d2f914c4f8568637544596d3
image: `perconalab/percona-xtradb-cluster-operator:PR-851-1378af4d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| storage | passed |
| haproxy | passed |
| scaling | passed |
| recreate | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | failed |
| self-healing-advanced | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| smart-update | passed |
| operator-self-healing | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| scheduled-backup | failed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/55a31bb44c890fafb0d92f834977d640996aa496
image: `perconalab/percona-xtradb-cluster-operator:PR-851-55a31bb4`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| storage | passed |
| recreate | passed |
| scaling | passed |
| haproxy | passed |
| big-data | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | failed |
| monitoring-2-0 | failed |
| self-healing-advanced | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| scheduled-backup | failed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/a15a30c918070f27f334e0508fa3526b6bb5e637
image: `perconalab/percona-xtradb-cluster-operator:PR-851-a15a30c9`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | passed |
| recreate | passed |
| storage | passed |
| haproxy | passed |
| init-deploy | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| demand-backup | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| pitr | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/ee88b4645caa51edbda62046e62810ef1ba8f8ee
image: `perconalab/percona-xtradb-cluster-operator:PR-851-ee88b464`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| scheduled-backup | passed |
| storage | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| scaling | passed |
| recreate | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| pitr | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/ee88b4645caa51edbda62046e62810ef1ba8f8ee
image: `perconalab/percona-xtradb-cluster-operator:PR-851-ee88b464`

**Review by delgod — DISMISSED** [review]

**Review by qJkee — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| recreate | passed |
| scheduled-backup | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| storage | passed |
| big-data | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing-advanced | passed |
| limits | passed |
| demand-backup | passed |
| smart-update | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |
| upgrade-consistency | passed |
| tls-issue-self | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/af10fbb5289f4b4cdf031e7421b526c8f73bc918
image: `perconalab/percona-xtradb-cluster-operator:PR-851-af10fbb5`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| haproxy | passed |
| pitr | passed |
| init-deploy | passed |
| demand-backup-encrypted-with-tls | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/851/commits/af10fbb5289f4b4cdf031e7421b526c8f73bc918
image: `perconalab/percona-xtradb-cluster-operator:PR-851-af10fbb5`

**Pull request merged**